components/private_channel/rust/ffi/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+#![allow(clippy::missing_safety_doc)]
+extern crate brave_private_channel;
+
+use std::convert::TryInto;
+use std::ffi::CStr;
+use std::os::raw::{c_char, c_int};
+use std::slice;
+
+pub const KEY_SIZE: usize = 32;
+
+#[repr(C)]
+pub struct ResultChallenge {
+    pub pkey_ptr: *const u8,
+    pub skey_ptr: *const u8,
+    pub key_size: usize,
+    pub shared_pubkey_ptr: *const u8,
+    pub encrypted_hashes_ptr: *const u8,
+    pub encrypted_hashes_size: usize,
+    pub error: bool,
+}
+
+impl Default for ResultChallenge {
+    fn default() -> ResultChallenge {
+        let mock_vec = vec![];
+        ResultChallenge {
+            error: true,
+            pkey_ptr: mock_vec.as_ptr(),
+            skey_ptr: mock_vec.as_ptr(),
+            key_size: 0,
+            shared_pubkey_ptr: mock_vec.as_ptr(),
+            encrypted_hashes_size: 0,
+            encrypted_hashes_ptr: mock_vec.as_ptr(),
+        }
+    }
+}
+
+#[repr(C)]
+pub struct ResultSecondRound {
+    pub encoded_partial_dec_ptr: *const u8,
+    pub encoded_partial_dec_size: usize,
+    pub encoded_proofs_ptr: *const u8,
+    pub encoded_proofs_size: usize,
+    pub random_vec_ptr: *const u8,
+    pub random_vec_size: usize,
+    pub error: bool,
+}
+
+impl Default for ResultSecondRound {
+    fn default() -> ResultSecondRound {
+        let mock_vec = vec![];
+        ResultSecondRound {
+            error: true,
+            encoded_partial_dec_ptr: mock_vec.as_ptr(),
+            encoded_partial_dec_size: 0,
+            encoded_proofs_ptr: mock_vec.as_ptr(),
+            encoded_proofs_size: 0,
+            random_vec_ptr: mock_vec.as_ptr(),
+            random_vec_size: 0,
+        }
+    }
+}
+
+/// Starts client attestation challenge;
+#[no_mangle]
+pub unsafe extern "C" fn client_start_challenge(
+    input: *const *const c_char,
+    input_size: c_int,
+    server_pk_encoded: *const u8,
+) -> ResultChallenge {
+    assert!(!input.is_null(), "Null pointers passed as input");
+    assert!(!server_pk_encoded.is_null(), "Null pointers passed as input");
```

Comment:
I removed `panic = "abort` from both profiles. I believe that will solve this point, let me know if I misunderstood and if something else is required @isislovecruft, thanks!

Reply:
Oh sorry, to be more clear, my recommendation is to keep `panic = "abort"` but not to _rely_ on it. So instead where you're writing code which could panic, to instead check the condition and manually abort, e.g. rather than:

```rust
assert!(!server_pk_encoded.is_null(), "Null pointers passed as input");
```

to do instead:

```rust
if server_pk_encoded.is_null() {
    ::std::process::abort();
}
```

Reply:
Ah, gotcha! Changing it now, thanks!

Reply:
- [ ] add `panic = "abort"` back and manually abort instead of using `assert`